# VIF and Multicollinearity

- VIF: Variance Inflation Factor measure how much the variance of an estimated regression coefficient is increased because of correlation among predictor variables in a model

- Multicollinearity: This happens when independent variables in a regression model are linearly correlated with each other. It does not reduce overall predictive power, but it makes coefficient estimates unstable and standard errors large.


How does VIF work?
- It calculates an R² value by regressing a single predictor variable against all other predictor variables in the model. The formula is:

$$
VIF = \frac{1}{1 - R^2}
$$


Interpreting the VIF Rules

- VIF = 1: The predictor does not correlate with other variables.
- VIF between 1 and 5: Low or moderate correlation, which is generally acceptable.
- VIF between 5 and 10: Moderate to high correlation that may warrant caution.
- VIF > 10: High multicollinearity that severely inflates standard errors and often requires action.

# How to fix multicollinearity

- To drop the columns
- Combine variables
- Add data
- Use alternative method: PCA (Principal Component Analysis) or Regularized Regression (Lasso, Ridge)

# Implementation using the library

In [1]:
import pandas as pd
from statsmodels.stats.outliers_influence import variance_inflation_factor 

In [2]:
# Example dataset
data = {
    "area_sqft": [1000, 1200, 1500, 1800, 2000, 2200, 2500, 2800],
    "bedrooms":  [2, 2, 3, 3, 4, 4, 5, 5],
    "house_size": [980, 1180, 1480, 1790, 1980, 2180, 2480, 2780],
    "age":       [20, 15, 10, 12, 8, 5, 3, 2]
}


In [3]:
df = pd.DataFrame(data)

In [4]:
df.head()

,area_sqft,bedrooms,house_size,age
0,1000,2,980,20
1,1200,2,1180,15
2,1500,3,1480,10
3,1800,3,1790,12
4,2000,4,1980,8


In [5]:
#Calculate the VIF
vif = pd.DataFrame()
vif["feature"] = df.columns
vif["VIF"] = [
    variance_inflation_factor(df.values, i)
    for i in range(df.shape[1])
]

print(vif)

      feature            VIF
0   area_sqft  380310.438961
1    bedrooms     323.319220
2  house_size  366049.038487
3         age       6.401138


# So, area_sqft and house_size has extremely high multicollinearity. We simply dropped the columns to remove the multi-collinearity


In [7]:
#So, removing the one columns
df = df.drop('house_size', axis = 1)
vif = pd.DataFrame()
vif['feature'] = df.columns
vif["VIF"] = [
    variance_inflation_factor(df.values, i)
    for i in range(df.shape[1])
]

print(vif)

     feature         VIF
0  area_sqft  223.941801
1   bedrooms  221.408575
2        age    1.790608


In [8]:
print(df[['area_sqft','bedrooms','age']].corr())

           area_sqft  bedrooms       age
area_sqft   1.000000  0.975195 -0.957798
bedrooms    0.975195  1.000000 -0.956399
age        -0.957798 -0.956399  1.000000


In [9]:
df_reduced = df.drop("bedrooms", axis=1)

In [10]:
X = df_reduced[["area_sqft", "age"]]

vif = pd.DataFrame()
vif["feature"] = X.columns
vif["VIF"] = [
    variance_inflation_factor(X.values, i)
    for i in range(X.shape[1])
]

print(vif)

     feature       VIF
0  area_sqft  1.782683
1        age  1.782683


# Final Results

Initially, the model showed severe multicollinearity, with very high VIF values for area_sqft, bedrooms, and house_size. After removing house_size, area_sqft and bedrooms still had very high VIF values, indicating strong multicollinearity between these variables. bedrooms was subsequently removed. The final VIF values for area_sqft and age were approximately 1.78, indicating that multicollinearity was no longer a significant issue.


So, The Final variables could fit:

- area_sqft
- age